### DISCUSSION TITLE
SSG and fallbacks

### DISCUSSION THREAD

**Discussion by jsonjunkie** [vercel/next.js#12500] [Help]
source code: https://github.com/JSONJunkie/nextjs_fallback
live: https://nextjs-fallback.herokuapp.com/

EDIT: New questions (but I'd appreciate any insight including answers to old questions): is it possible to serve/trigger fallback strictly based on the available/rendered pages/paths instead of based on what's going on inside getStaticProps (or what getStaticProps fetches)? Also is there some method on router object that exposes available/rendered pages/paths?

Hello, 

Looking to get some insight to the use case I am trying to achieve here. I am trying to have client-side navigation from index page and make use of fallbacks for SSG dynamic pages/routes that weren't pre rendered during build time. 

Currently, I can only get my fallback to render for these ungenerated pages/routes if the route is typed directly into the address bar or the page refreshes during navigation (I can achieve this by incorrectly using next's Link component as shown in the code I linked). If I try to navigate to a page/route that hasn't been generated yet correctly for client side navigation, nothing happens until getStaticProps finishes. The fallback is never rendered and the content is shown for the newly generated page after getStaticProps finishes with no refresh.

So is it possible to route to a dynamic page that hasn't been generated yet using the Link component correctly for client-side navigation and have the fallback render too without having the page refresh? Thanks in advance for any help!

**Comment by brandonpittman** [comment]
@JSONJunkie Same issue here, I believe. My /posts page has cards with post links that also have tag links inside each card. The dynamic tag links are trying to fetch JSON that isn't there yet, and then trying to visit the dynamic route throws an error as well.

**Comment by jaybytez** [comment]
I was thinking of creating my own question, but didn't want to duplicate this. Also, don't want to derail this question from getting answered.

We are running a test with SSG - Dynamic Links - Fallback=true. What is interesting is we have a Component and the StaticPaths and StaticProps functions.  When you hit a unique URL, it doesn't hit StaticPaths or StaticProps, it hits the Component first.  The object passed to Component is empty and therefore the page errors (we bypass this by creating a "Default" version of the object if it is null).  So by supplying the empty object, we bypass this, but we see the page actually refresh twice.

So SSG - Dynamic Links - Fallback=true is not working initially for us.

Hoping Next team can give insight to this.

**Comment by jsonjunkie** [comment]
@brandonpittman Yea our issues sound similar. Something going on with dynamic routing and getStaticProps for me I think.

@jaybytez I believe that's what was happening for me too, that when navigating to a unique URL, it hits the component first instead of staticprops/staticpaths. With fallback=false, I was able to get a render that should have resulted in a 404 last I tried. 

I've kinda brute forced a solution for myself but I am not sure at all if my use of getStaticProps is really "legal" or legitimate. For a portfolio project, I'm building a language translator that generates static html pages after hitting a translator api that contains the requested translation. The static html will be served upon future requests instead of having to go through the trouble of hitting the translator api again which makes the navigation super fast. What I've done is store the request in a database and return null from getstaticprops if the translation is not found which results in a 404 because the translation hasn't been generated yet and then a TypeError when nextjs tries to render a component with null data which triggers a refresh/fallback where this time, requested translation will be generated in getstaticprops and properly rendered because the request now exists in the database from the previous failure of getstaticprops.

**Comment by jaybytez** [comment]
One of the things we just tried was in our component, testing if the object coming in was undefined.  If it exists, we execute normally and if it doesn't we return ''.  What this does is shows a blank page that is quickly refreshed with the actual page.  

So we have a page [,,,slugs].js and we initially know some of the paths and they are populated in static paths.  So a link like /item and then we test with a new dynamic link like /fr/item (this is getting the error).  Because it follows an execution path of our Component then getStaticPaths then our Component.  

So we did this:
`export default function Component({item}) {
  console.log('Component - Enter');

  if (item !== undefined) {
    return (
		&lt;h1&gt;{ item.name }&lt;/h1&gt;
    );
  } else {
    return '';
  }
}`

------ Updating ------
This seemed to help, the it still remains that the page refreshes twice.  The first hit has no data and the second hit does.

**Comment by jsonjunkie** [comment]
Are you guys rendering a fallback component? I was doing something similar to this previously where I added this to my if fallback statement to show the fallback if the object coming in was undefined:

if (router.isFallback || !data) 

I tried to return " as well just to be sure but in both of these cases, while I don't get any errors anymore when navigating to a page/translation that doesn't exist, getStaticProps also doesn't run again before generating the static html page in both dev and production environments (I think because I don't get any errors anymore?) so it generates static html right after the request is stored in the database but before the translation is made which results in static html without the translation and the translator api is never hit because the translation happens when getStaticProp runs for the second time after failing to render a component (due to the null/undefined object being returned from the first getStaticProps run). 

I also tried to "cheat" by navigating to a translaton/page that has been requested but not yet translated and when I return "" if the object returned is undefined, I do see a refresh and eventually the desired static page with the translation but in this case, I never see the fallback component rendered unless i explicitly return it also when the object returned is undefined: 

if (router.isFallback || !data) 

When returning "", I can see that router.isFallback is true (I have a backdrop that only shows when router.isFallback is true when navigating) but because I return "", I think it just stays on the blank page instead of ever rendering the fallback until getStaticProps finishes and the static page is ready to be rendered. This is something I didn't realize initially when I made this post and I'm not sure how to force the fallback to render when returning "" if the object is undefined from getStaticProps because the blank page is being rendered until getStaticProps is finished. If I return the fallback component if the object is undefined from getStaticProps, then I get the fallback component until getStaticProps is finished as desired.

It's like my case relies on the (kinda?) recursive nature of getStaticPaths that only results when it fails to render a component and gives a TypeError from the first time getStaticProps runs
